### PR TITLE
Add coin per group limit.

### PR DESCRIPTION
**Pull request by thebevrishot**
push mints on new block to next group if current_coins + new_coins > 16000

### CODE REVIEW

File: src/zerocoin_v3.cpp [hunk 1]
```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
...
+
+        // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
+        // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
+        // each block will end up in the same accumulator.
+        SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
+        int coinsPerId = ZC_SPEND_V3_COINSPERID;
+        int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
+        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
+            && coinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit)
+            || coinGroup.lastBlock == index // or we have already placed some coins from current block.
```
Comment (ultimaweapon) [inline src/zerocoin_v3.cpp]: Seem like this one still allow mint per group to exceed limit.

File: src/zerocoin_v3.cpp [hunk 2]
```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
...
-        newCoinGroup.firstBlock = newCoinGroup.lastBlock = index;
-        newCoinGroup.nCoins = 1;
-    }
-    CMintedCoinInfo coinInfo;
-    coinInfo.denomination = denomination;
-    coinInfo.id = mintCoinGroupId;
-    coinInfo.nHeight = index->nHeight;
-    mintedPubCoins.insert(std::make_pair(pubCoin, coinInfo));
-    return mintCoinGroupId;
+    return 1;
```
Comment (martun) [inline src/zerocoin_v3.cpp]: return true?

File: src/zerocoin_v3.cpp [hunk 3]
```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
...
+            latestCoinIds[denomination] = 1;
+        int	mintCoinGroupId = latestCoinIds[denomination];
+
+        // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
+        // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
+        // each block will end up in the same accumulator.
+        SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
+        int coinsPerId = ZC_SPEND_V3_COINSPERID;
+        int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
+        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
```
Comment (martun) [inline src/zerocoin_v3.cpp]: Why is the first part of "if" necessary? Check in "oinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit" checks that all the mints in this block will fit in the same accumulator right? And I believe there must be coinsPerId used, not coinsPerIdLimit.

File: src/zerocoin_v3.cpp [hunk 4]
```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
...
+            && coinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit)
```
Comment (levonpetrosyan93) [inline src/zerocoin_v3.cpp]: I think better to do this check for each denomination, not for each coin.

File: src/zerocoin_v3.cpp (hunk 1 above)
Comment (levonpetrosyan93) [reply src/zerocoin_v3.cpp]: Agree with @ultimaweapon.

File: src/zerocoin_v3.cpp (hunk 2 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: you're right. thanks.

File: src/zerocoin_v3.cpp (hunk 4 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: good idea. thanks.

File: src/zerocoin_v3.cpp (hunk 3 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: I use 15 000 as soft cap(`coinsPerID`) and 16 000(`coinsPerIdLimit`) as hard cap. 

If before add new mints coin in group is greater than or equal 15000, then new mints should be added to next group. which is first case. `coinGroup.nCoins < coinsPerId`

and if mints in group is less than 15000 but new mints + old mints > 16000, then new mints should also be push to new group. which is second case `coinGroup.nCoins + new mints amount < coinsPerIdLimit`

File: src/zerocoin_v3.cpp (hunk 1 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: quite not understand how it allow but have a bug here I have fixed. please check.

File: src/zerocoin_v3.cpp [hunk 5]
```diff
@@ -600,29 +600,26 @@ bool CSigmaState::AddMintsToStateAndBlockIndex(
...
         SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
         int coinsPerId = ZC_SPEND_V3_COINSPERID;
         int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
-        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
-            && coinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit)
-            || coinGroup.lastBlock == index // or we have already placed some coins from current block.
-            ) {
-            if (coinGroup.nCoins++ == 0) {
+        if (coinGroup.nCoins < coinsPerId // there's still space in the accumulator
+            && coinGroup.nCoins + mintsWithThisDenom.size() <= coinsPerIdLimit) {
```
Comment (ultimaweapon) [inline src/zerocoin_v3.cpp]: Seem like this one will cause signesss comparison mismatch.

File: src/zerocoin_v3.cpp [hunk 6]
```diff
@@ -637,23 +634,27 @@ bool CSigmaState::AddMintsToStateAndBlockIndex(
...
+            coinInfo.id = mintCoinGroupId;
+            coinInfo.nHeight = index->nHeight;
+            mintedPubCoins.insert(std::make_pair(mint, coinInfo));
+
+            LogPrintf("AddMintsToStateAndBlockIndex: mint added denomination=%d, id=%d\n", denomination, mintCoinGroupId);
+            index->sigmaMintedPubCoins[{denomination, mintCoinGroupId}].push_back(mint);
+        }
     }
-    return 1;
+    return true;
```
Comment (ultimaweapon) [inline src/zerocoin_v3.cpp]: Seem like we don't need a returned value anymore. So change this method to `void` instead will be better.

**Comment by thebevrishot** [comment]
@ultimaweapon some thing like `coinGroup.lastBlock->nHeight <= index->nHeight` ?

File: src/zerocoin_v3.cpp (hunk 6 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: make sense.

File: src/zerocoin_v3.cpp (hunk 5 above)
Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: fixed

File: src/zerocoin_v3.cpp (hunk 3 above)
Comment (martun) [reply src/zerocoin_v3.cpp]: Wait, I did not understand why we need 2 caps any more. We must do the following
1. Know how many coins are already in the last accumulator.
2. Count how many new mints are in the new block.
3. If these mints + number of coins already in the accumulator < 16.000, then put them all into the same accumulator, if not, then close the current accumulator and put them all into the next accumulator. 

This must be the whole logic, no soft cap.

Comment (thebevrishot) [reply src/zerocoin_v3.cpp]: I see, that's make sense.

**Pull request merged**